anitya/lib/backends/github.py

```diff
@@ -59,18 +65,168 @@ def get_versions(cls, project):
+
+        if 'errors' in json:
+            error_str = ''
+            for error in json['errors']:
+                error_str += '"%s": "%s"\n' % (
+                    error['type'], error['message'])
+            raise AnityaPluginException(
+                '%s: Server responded with following errors\n%s' % (
+                    project.name, error_str))
+
+        total_count = json['data']['repository']['refs']['totalCount']
+
+        _log.debug('Received %s tags for %s' % (total_count, project.name))
+
+        # TODO: Save cursor to project and use it to fetch only newer versions
+        # cursor = json['data']['repository']['refs']['edges']['cursor']
+
+        remaining = json['data']['rateLimit']['remaining']
+        reset_time = json['data']['rateLimit']['resetAt']
+        _log.debug('Github API ratelimit remains %s, will reset at %s UTC' % (
+                remaining, reset_time))
+
+        versions = []
+
+        for edge in json['data']['repository']['refs']['edges']:
+            version = edge['node']['name']
+            if project.version_prefix is not None and \
+                    version.startswith(project.version_prefix):
+                version = version[len(project.version_prefix):]
+            elif version.startswith('v'):
+                version = version[1:]
+            # TODO: Save commit url to DB and make versions clickable on frontend
+            # commit_url = edge['node']['target']['commitUrl']
+            versions.append(version)
+
+        return versions
+
+    @classmethod
+    def prepare_query(cls, owner, repo, after=''):
+        ''' Method for preparing GraphQL query for specified repository
+
+        :arg owner: owner of the repository
+        :type owner: str
+        :arg repo: repository name
+        :type repo: str
+        :arg after: cursor id of the latest received commit
+        :type after: str
+        :return: GraphQL query
+        :return type: str
+        '''
+
+        after_str = ''
+
+        # Fill cursor if we have the id
+        if after:
+            after_str = ', after: "%s"' % after
+
+        query = '''
+        {
+            repository(owner: "%s", name: "%s") {
+                refs (refPrefix: "refs/tags/", last:50,
+                        orderBy:{field:TAG_COMMIT_DATE, direction:ASC}%s) {
+                    totalCount
+                    edges {
+                        cursor
+                        node {
+                            name
+                            target {
+                                commitUrl
+                            }
+                        }
+                    }
+                }
+            }
+            rateLimit {
+                limit
+                remaining
+                resetAt
+            }
+        }
+        ''' % (owner, repo, after_str)
+
+        return query
+
+    @classmethod
+    def call_url_post(cls, url, json, insecure=False, use_token=False):
```

Comment:
I think you can just factor this into the ``get_versions`` call. I think it's reasonable for the token to be required, and we should never use ``insecure=True`` with GitHub so it boils down to a single call to another function.

The ``get_headers`` interface (while matching the existing style) is more complicated than just setting a default set of headers somewhere, then making a copy here and injecting the access token. I recommend doing that.

Reply:
I didn't understood the second paragraph in your comment. You want me to copy the `get_headers` method to this backend.
The `get_headers` method is new method in `BaseBackend` and it only fills basic header fields.

Reply:
Ah, what I'm saying is I don't like the use of inheritance in the backends. It's complicated and adds no value, but I never got a chance to refactor it. Rather than add to it, this could be implemented as:

In constants or BaseBackend's module or something:
```
REQUEST_HEADERS = <default_headers>
```

Then for your Github backend you just make a copy of REQUEST_HEADERS and add your auth token:

```
github_headers = REQUEST_HEADERS.copy()
github_headers['Authentication'] = ...
```

There's no need for a get_headers method that subclasses override. The whole ``call_url`` function has no business being a class method at all. If should just be a module-level function that acts as a wrapper that handles the special FTP case and otherwise is the requests API.

Reply:
I thought that the inheritance is intentional, otherwise `BaseBackend` doesn't have much sense.

Reply:
Inheritance is fine to define the interface, but the only interface we're really interested in is for a given project, new versions are returned. Everything else is an implementation detail of the particular backend and doesn't need to be part of the class.